### PR TITLE
Removing credsSecret required field in tenant spec

### PR DESCRIPTION
**Pull request by Alevsk**
- Removing required credsSecret field in tenant spec
- Migration from creddsSecret to Configuration field for old tenants

Signed-off-by: Lenin Alevski <alevsk.8772@gmail.com>

### CODE REVIEW

File: pkg/controller/cluster/main-controller.go [hunk 1]
```diff
@@ -601,6 +601,11 @@ func (c *Controller) syncHandler(key string) error {
 
 	tenant.EnsureDefaults()
 
+	// Check the Sync Version to see if the tenant needs upgrade
+	if tenant, err = c.checkForUpgrades(ctx, tenant); err != nil {
+		return err
+	}
+
```
Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:608]: Do not check for upgrades without validating the tenant spec. @Alevsk

File: pkg/controller/cluster/upgrades.go [hunk 2]
```diff
@@ -309,3 +317,64 @@ func (c *Controller) upgrade430(ctx context.Context, tenant *miniov2.Tenant) (*m
...
+			_, err = c.minioClientSet.MinioV2().Tenants(tenant.Namespace).Update(ctx, tenantCopy, metav1.UpdateOptions{})
+			if err != nil {
+				klog.Errorf("Error updating tenant '%s/%s', could not update tenant.spec.configuration field: %v", tenant.Namespace, tenant.Name, err)
+			}
+		} else {
+			return nil, fmt.Errorf("'%s/%s' Error migrating tenant credsSecret, credsSecret not present", tenant.Namespace, tenant.Name)
+		}
+	}
+	return c.updateTenantSyncVersion(ctx, tenant, version447)
+}
```
Comment (harshavardhana) [inline pkg/controller/cluster/upgrades.go:380]: We should perhaps do this for v4.4.8 -made a new release to make sure to fix an important bug during deployment.

File: examples/kustomization/tenant-test/tenantNamePatch.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+- op: replace
+  path: /metadata/name
+  value: storage-test
```
Comment (harshavardhana) [inline examples/kustomization/tenant-test/tenantNamePatch.yaml:3]: add a missing newline

File: examples/kustomization/tenant-test/tenant-credssecret.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+data:
+  accesskey: bWluaW9hZG1pbg==
+  secretkey: bWluaW9hZG1pbg==
+kind: Secret
+metadata:
+  name: storage-credssecret
+type: Opaque
```
Comment (harshavardhana) [inline examples/kustomization/tenant-test/tenant-credssecret.yaml:8]: add a missing newline

File: examples/kustomization/tenant-test/kustomization.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+namespace: tenant-test
+patchesStrategicMerge:
+  - tenant.yaml
+patchesJson6902:
+  - target:
+      group: minio.min.io
+      version: v2
+      kind: Tenant
+      name: storage
+    path: tenantNamePatch.yaml
```
Comment (harshavardhana) [inline examples/kustomization/tenant-test/kustomization.yaml:15]: add a missing newline

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — DISMISSED** [review]

**Review by harshavardhana — APPROVED** [review]

File: kubectl-minio/cmd/resources/secret.go [hunk 6]
```diff
@@ -17,16 +17,20 @@ package resources
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 func tenantSecretData() map[string][]byte {
-	m := make(map[string][]byte, 2)
-	m["accesskey"] = []byte(uuid.New().String())
-	m["secretkey"] = []byte(uuid.New().String())
-	return m
+	m := map[string]string{}
+	m["MINIO_BROWSER"] = "on"
```
Comment (harshavardhana) [inline kubectl-minio/cmd/resources/secret.go:28]:
```suggestion
```

This is always `on` by default no need to explicitly set this.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
PTAL @Alevsk at conflicts

**Comment by harshavardhana** [comment]
Can we rebase this and take it in @Alevsk ?

**Comment by Alevsk** [comment]
> Can we rebase this and take it in @Alevsk ?

Working on this today @harshavardhana

File: pkg/controller/cluster/upgrades.go [hunk 7]
```diff
@@ -296,3 +302,63 @@ func (c *Controller) upgrade430(ctx context.Context, tenant *miniov2.Tenant) (*m
...
+		accessKey = string(credsSecret.Data["accesskey"])
+	}
+	if _, ok := credsSecret.Data["secretkey"]; ok {
+		secretKey = string(credsSecret.Data["secretkey"])
+	}
+	if accessKey == "" || secretKey == "" {
+		return nil, fmt.Errorf("'%s/%s' Error migrating tenant credsSecret to Configuration", tenant.Namespace, tenant.Name)
+	}
+	tenantConfiguration := map[string]string{}
+	tenantConfiguration["MINIO_BROWSER"] = "on"
```
Comment (harshavardhana) [inline pkg/controller/cluster/upgrades.go:332]:
```suggestion
```
This is default doesn't need to be set explicitly.

File: pkg/controller/cluster/upgrades.go [hunk 8]
```diff
@@ -296,3 +302,63 @@ func (c *Controller) upgrade430(ctx context.Context, tenant *miniov2.Tenant) (*m
...
+	var accessKey string
+	var secretKey string
+	if _, ok := credsSecret.Data["accesskey"]; ok {
+		accessKey = string(credsSecret.Data["accesskey"])
+	}
+	if _, ok := credsSecret.Data["secretkey"]; ok {
+		secretKey = string(credsSecret.Data["secretkey"])
+	}
+	if accessKey == "" || secretKey == "" {
+		return nil, fmt.Errorf("'%s/%s' Error migrating tenant credsSecret to Configuration", tenant.Namespace, tenant.Name)
```
Comment (harshavardhana) [inline pkg/controller/cluster/upgrades.go:329]:
```suggestion
		return nil, fmt.Errorf("accessKey/secretKey are empty - '%s/%s' error in migrating tenant credsSecret to newer configuration", tenant.Namespace, tenant.Name)
```

File: pkg/controller/cluster/upgrades.go [hunk 9]
```diff
@@ -296,3 +302,63 @@ func (c *Controller) upgrade430(ctx context.Context, tenant *miniov2.Tenant) (*m
...
 }
+
+// Upgrades the sync version to v4.4.26
+// in this version we finally deprecated tenant.spec.credsSecret field.
+func (c *Controller) upgrade4426(ctx context.Context, tenant *miniov2.Tenant) (*miniov2.Tenant, error) {
+	if tenant.HasConfigurationSecret() {
+		return c.updateTenantSyncVersion(ctx, tenant, version4426)
+	}
+	if !tenant.HasCredsSecret() {
+		return nil, fmt.Errorf("'%s/%s' Error migrating tenant credsSecret, credsSecret not present", tenant.Namespace, tenant.Name)
```
Comment (harshavardhana) [inline pkg/controller/cluster/upgrades.go:313]:
```suggestion
		return nil, fmt.Errorf("'%s/%s' error migrating tenant credsSecret, credsSecret does not exist", tenant.Namespace, tenant.Name)
```

File: pkg/controller/cluster/upgrades.go [hunk 10]
```diff
@@ -296,3 +302,63 @@ func (c *Controller) upgrade430(ctx context.Context, tenant *miniov2.Tenant) (*m
...
+	// Update tenant fields
+	tenantCopy := tenant.DeepCopy()
+	tenantCopy.EnsureDefaults()
+	tenantCopy.Spec.Configuration = &corev1.LocalObjectReference{
+		Name: tenantCopy.ConfigurationSecretName(),
+	}
+	tenantCopy.Spec.CredsSecret = nil
+	_, err = c.minioClientSet.MinioV2().Tenants(tenant.Namespace).Update(ctx, tenantCopy, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("Error updating tenant '%s/%s', could not update tenant.spec.configuration field: %v", tenant.Namespace, tenant.Name, err)
```
Comment (harshavardhana) [inline pkg/controller/cluster/upgrades.go:361]: return error here

File: pkg/controller/cluster/main-controller.go [hunk 11]
```diff
@@ -745,6 +746,10 @@ func (c *Controller) syncHandler(key string) error {
 
 	adminClnt, err := tenant.NewMinIOAdmin(tenantConfiguration, c.getTransport())
 	if err != nil {
+		if _, err = c.updateTenantStatus(ctx, tenant, StatusTenantCredentialsNotSet, 0); err != nil {
+			return err
+		}
+		klog.Errorf("Error instantiating adminClnt: %v", err)
```
Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:752]:
```suggestion
		klog.Errorf("Error initializing minio admin client: %v", err)
```

File: pkg/controller/cluster/main-controller.go [hunk 12]
```diff
@@ -745,6 +746,10 @@ func (c *Controller) syncHandler(key string) error {
 
 	adminClnt, err := tenant.NewMinIOAdmin(tenantConfiguration, c.getTransport())
 	if err != nil {
+		if _, err = c.updateTenantStatus(ctx, tenant, StatusTenantCredentialsNotSet, 0); err != nil {
```
Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:749]:
```suggestion
		if _, err = c.updateTenantStatus(ctx, tenant, StatusTenantCredentialsNotSet, 0); err != nil {
```
This error shadows the error returned by `tenant.NewMinIOAdmin()` leading to the err being `nil` 

Instead you should use a newer variable such as `uerr` or something.

Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:749]:
```suggestion
		if _, uerr := c.updateTenantStatus(ctx, tenant, StatusTenantCredentialsNotSet, 0); uerr != nil {
```

File: pkg/controller/cluster/main-controller.go [hunk 13]
```diff
@@ -745,6 +746,10 @@ func (c *Controller) syncHandler(key string) error {
...
+			return err
```
Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:750]:
```suggestion
			return uerr
```

File: pkg/controller/cluster/main-controller.go [hunk 14]
```diff
@@ -111,6 +111,7 @@ const (
 	StatusUpdatingAffinity                   = "Updating Pod Affinity"
 	StatusNotOwned                           = "Statefulset not controlled by operator"
 	StatusFailedAlreadyExists                = "Another MinIO Tenant already exists in the namespace"
+	StatusTenantCredentialsNotSet            = "Credentials for tenant not properly set"
```
Comment (harshavardhana) [inline pkg/controller/cluster/main-controller.go:114]:
```suggestion
	StatusTenantCredentialsNotSet            = "Tenant credentials are not set properly"
```

File: pkg/apis/minio.min.io/v2/helper.go [hunk 15]
```diff
@@ -897,8 +897,8 @@ func (t *Tenant) Validate() error {
 		return errors.New("pools must be configured")
 	}
 
-	if t.Spec.CredsSecret == nil {
-		return errors.New("please set credsSecret secret with credentials for Tenant")
+	if !t.HasConfigurationSecret() && !t.HasCredsSecret() {
+		return errors.New("please set Configuration secret with credentials for Tenant")
```
Comment (harshavardhana) [inline pkg/apis/minio.min.io/v2/helper.go:901]:
```suggestion
		return errors.New("please set 'configuration' secret with credentials for Tenant")
```

File: pkg/apis/minio.min.io/v1/helper.go [hunk 16]
```diff
@@ -632,8 +626,8 @@ func (t *Tenant) Validate() error {
 		return errors.New("zones must be configured")
 	}
 
-	if t.Spec.CredsSecret == nil {
-		return errors.New("please set credsSecret secret with credentials for Tenant")
+	if t.Spec.Configuration == nil {
+		return errors.New("please set configuration secret with credentials for Tenant")
```
Comment (harshavardhana) [inline pkg/apis/minio.min.io/v1/helper.go:630]:
```suggestion
		return errors.New("please set 'configuration' secret with credentials for Tenant")
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
LGTM overall, few comments

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
We should take this perhaps before v4.4.26 @dvaldivia @cniackz ?

**Comment by harshavardhana** [comment]
>  We should take this perhaps before v4.4.26 @dvaldivia @cniackz ?

ping @cniackz @dvaldivia

**Review by harshavardhana — APPROVED** [review]

**Comment by cniackz** [comment]
@Alevsk please review conflicts when you have a chance.

**Review by cniackz — COMMENTED** [review]
@Alevsk please resolve the conflicts 👍

**Comment by Alevsk** [comment]
> I'm hitting an NPE with a tenant which has no config file
> 
> ```
> goroutine 361 [running]:
> k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x4dd574})
>     k8s.io/apimachinery@v0.20.6/pkg/util/runtime/runtime.go:55 +0xd8
> panic({0x16ad080, 0x2759220})
>     runtime/panic.go:1038 +0x215
> k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000294250})
>     k8s.io/apimachinery@v0.20.6/pkg/util/runtime/runtime.go:55 +0xd8
> panic({0x16ad080, 0x2759220})
>     runtime/panic.go:1038 +0x215
> github.com/minio/operator/pkg/controller/cluster.(*Controller).checkForUpgrades(0xc000450480, {0x1af9218, 0xc000042028}, 0xc000e5f180)
>     github.com/minio/operator/pkg/controller/cluster/upgrades.go:93 +0xaca
> github.com/minio/operator/pkg/controller/cluster.(*Controller).syncHandler(0xc000450480, {0xc001235bc0, 0x17})
>     github.com/minio/operator/pkg/controller/cluster/main-controller.go:650 +0x5ab
> github.com/minio/operator/pkg/controller/cluster.(*Controller).processNextWorkItem.func1({0x163e060, 0xc000294250})
>     github.com/minio/operator/pkg/controller/cluster/main-controller.go:560 +0x245
> github.com/minio/operator/pkg/controller/cluster.(*Controller).processNextWorkItem(0xc000450480)
>     github.com/minio/operator/pkg/controller/cluster/main-controller.go:572 +0x62
> github.com/minio/operator/pkg/controller/cluster.(*Controller).runWorker(0x0)
>     github.com/minio/operator/pkg/controller/cluster/main-controller.go:512 +0x70
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7fe77453cd60)
>     k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:155 +0x67
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0003e90e0, {0x1acbfc0, 0xc0006f4e70}, 0x1, 0xc00008f620)
>     k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:156 +0xb6
> k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000998520, 0x3b9aca00, 0x0, 0x28, 0xc0003e9140)
>     k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:133 +0x89
> k8s.io/apimachinery/pkg/util/wait.Until(0xc000998590, 0xc0001c6480, 0xc00099c7e0)
>     k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:90 +0x25
> created by github.com/minio/operator/pkg/controller/cluster.(*Controller).Start.func1
>     github.com/minio/operator/pkg/controller/cluster/main-controller.go:397 +0x3c5
> Stream closed EOF for minio-operator/minio-operator-6f4f49887d-55pj4 (minio-operator)
> ```

If tenant has no config file we should throw an error and stop execution for that tenant, meaning if theres a version of that tenant already running thats fine,  but we should not apply any new incoming changes, that sounds good?

**Review by dvaldivia — APPROVED** [review]

**Comment by harshavardhana** [comment]
Are we ready to make v5.0.0 @dvaldivia ?

**Pull request merged**